10-Merge/readme.md

```diff
@@ -0,0 +1,202 @@
+	$ cat file
+	.
+	.
+	.
+	.
+	.
+
+The file `file` was changed in both branches `bar` and `foo` and we want to get those changes back into the master. We will start with `bar`
+
+	$ git merge bar
+	Merge made by the 'recursive' strategy.
+	 file | 4 ++--
+	 1 file changed, 2 insertions(+), 2 deletions(-)
+
+> Running git merge with non-trivial uncommitted changes is discouraged: while possible, it may leave you in a state that is hard to back out of in the case of a conflict.
+
+This command produces no conflicts, since no changes we made to `file` after branching into `bar`. So after this merge, we can see that repository looks like this:
+
+	$ git log --all --graph --oneline --decorate
+	*   05dbdf3 (HEAD -> master) Merge branch 'bar'
+	|\  
+	| * bfb5647 (origin/bar, bar) bar
+	* | c5d665f (origin/master, origin/HEAD) fixed readme a bit
+	| | * 4909911 (origin/cherry-pick, cherry-pick) merge conflict in file
+	| | * e7f0e30 added file for cherry_pick
+	| |/  
+	|/|   
+	* | 286a273 added readme
+	|/  
+	| * c5c3da1 (origin/foo, foo) file2 for demonstration of cherry-pick
+	| * c361f2e foo
+	|/  
+	* 25f7328 file
+
+	$ cat file
+	.
+	bar
+	.
+	.
+	bar
+
+We can see that branch `bar` was merged into `master`. Now it's time to do the same with `foo`.
+
+	$ git merge foo
+	Auto-merging file
+	CONFLICT (content): Merge conflict in file
+	Automatic merge failed; fix conflicts and then commit the result.
+
+We have a conflict! We need to somehow resolve it.
+
+> You can always abort merge with `git merge --abort`
+
+Text editor way
+---------------
+
+In case of conflict, git puts changes into a file like this:
+
+	$ cat file
+	.
+	<<<<<<< HEAD
+	bar
+	.
+	.
+	bar
+	=======
+	foo
+	.
+	foo
+	.
+	>>>>>>> foo
+
+The `=======` separates two versions while `HEAD` being current branch and `foo` the merged branch (`foo`). All you have to is to modified the file into what you want it to look like, so I'll do for example this:
+
+	$ cat file
+	.
+	foobar
+	.
+	foo
+	bar
+
+After you are done with manual merge, you need to mark conflict as resolved and commit the merge:
+
+	$ git add file
+	$ git commit -m "Merge branch 'foo'"
+	[master b702845] Merge branch 'foo'
+
+Let's check it out:
+
+	$ git log --all --graph --oneline --decorate
+	*   b702845 (HEAD -> master) Merge branch 'foo'
+	|\  
+	| * c5c3da1 (origin/foo, foo) file2 for demonstration of cherry-pick
+	| * c361f2e foo
+	* |   05dbdf3 Merge branch 'bar'
+	|\ \  
+	| * | bfb5647 (origin/bar, bar) bar
+	| |/  
+	* | c5d665f (origin/master, origin/HEAD) fixed readme a bit
+	| | * 4909911 (origin/cherry-pick, cherry-pick) merge conflict in file
+	| | * e7f0e30 added file for cherry_pick
+	| |/  
+	|/|   
+	* | 286a273 added readme
+	|/  
+	* 25f7328 file
+
+We now have merge both `bar` and `foo` into the master, in the process we resolved conflict between them.
+
+Mergetool way
+-------------
+
+We throw away the last commit (the merge of `foo`), so we can do it again, this time using merge tool.
```

Comment:
It would be useful to put here reference to the merge tools configuration and make a comment that it is a prereq